zcash_primitives/src/transaction/builder.rs

```diff
@@ -0,0 +1,692 @@
+        witness: IncrementalWitness<Node>,
+    ) -> Result<(), Error> {
+        // Consistency check: all anchors must equal the first one
+        if let Some(anchor) = self.anchor {
+            let witness_root: Fr = witness.root().into();
+            if witness_root != anchor {
+                return Err(Error(ErrorKind::AnchorMismatch));
+            }
+        } else {
+            self.anchor = Some(witness.root().into())
+        }
+        let witness = witness.path().ok_or(Error(ErrorKind::InvalidWitness))?;
+
+        let alpha = Fs::rand(&mut self.rng);
+
+        self.mtx.value_balance.0 += note.value as i64;
+
+        self.spends.push(SpendDescriptionInfo {
+            extsk,
+            diversifier,
+            note,
+            alpha,
+            witness,
+        });
+
+        Ok(())
+    }
+
+    /// Adds a Sapling address to send funds to.
+    pub fn add_sapling_output(
+        &mut self,
+        ovk: OutgoingViewingKey,
+        to: PaymentAddress<Bls12>,
+        value: Amount,
+        memo: Option<Memo>,
+    ) -> Result<(), Error> {
+        let output = SaplingOutput::new(&mut self.rng, ovk, to, value, memo)?;
+
+        self.mtx.value_balance.0 -= value.0;
+
+        self.outputs.push(output);
+
+        Ok(())
+    }
+
+    /// Adds a transparent address to send funds to.
+    pub fn add_transparent_output(
+        &mut self,
+        to: &TransparentAddress,
+        value: Amount,
+    ) -> Result<(), Error> {
+        if value.0 < 0 {
+            return Err(Error(ErrorKind::InvalidAmount));
+        }
+
+        self.mtx.vout.push(TxOut {
+            value,
+            script_pubkey: to.script(),
+        });
+
+        Ok(())
+    }
+
+    /// Sets the Sapling address to which any change will be sent.
+    ///
+    /// By default, change is sent to the Sapling address corresponding to the first note
+    /// being spent (i.e. the first call to [`Builder::add_sapling_spend`]).
+    pub fn send_change_to(&mut self, ovk: OutgoingViewingKey, to: PaymentAddress<Bls12>) {
+        self.change_address = Some((ovk, to));
+    }
+
+    /// Builds a transaction from the configured spends and outputs.
+    ///
+    /// Upon success, returns a tuple containing the final transaction, and the
+    /// [`TransactionMetadata`] generated during the build process.
+    ///
+    /// `consensus_branch_id` must be valid for the block height that this transaction is
+    /// targeting. An invalid `consensus_branch_id` will *not* result in an error from
+    /// this function, and instead will generate a transaction that will be rejected by
+    /// the network.
+    pub fn build(
+        mut self,
+        consensus_branch_id: u32,
+        prover: impl TxProver,
+    ) -> Result<(Transaction, TransactionMetadata), Error> {
+        let mut tx_metadata = TransactionMetadata::new();
+
+        //
+        // Consistency checks
+        //
+
+        // Valid change
+        let change = self.mtx.value_balance.0
```

Comment:
This also has the potential for underflows / overflows (the transparent outputs have arbitrarily chosen values as well.)